Set CMake to default to single instead of reentrant

albinahlback opened the pull request:
Solves #1392

fredrik-johansson commented:
This fails:

    2023-05-22T15:37:26.3851996Z fmpz.c.obj : error LNK2019: unresolved external symbol atomic_add_fetch referenced in function _fmpz_cleanup
    2023-05-22T15:37:26.3871912Z bin\flint-18.dll : fatal error LNK1120: 1 unresolved externals

albinahlback commented:
Weird. Shouldn't be related to this PR though, just that this triggers it. Perhaps we should have two different CIs for MSVC. I'll take a look in a month.

albinahlback commented:
Alright, looks like I have fixed it?

Before, the default was to use reentrant mode for CMake. Now, the default is set to single.

Here, `atomic_add_fetch` was not a function which was generated for MSVC. But, as we use C11, we can use atomics, that is, using `stdatomic.h`. With this, we could replace
```c
#if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8)) && FLINT_USES_PTHREAD
       new_count = __atomic_add_fetch(&(ptr->count), 1, __ATOMIC_SEQ_CST);
#elif defined(_MSC_VER) && FLINT_USES_PTHREAD
       new_count = atomic_add_fetch(&(ptr->count), 1);
#else /* may be a very small leak with pthreads */
       new_count = ++ptr->count;
#endif
```
with
```c
#if FLINT_USES_PTHREAD
       new_count = atomic_fetch_add(&(header_ptr->count), 1);
#else
       new_count = ++header_ptr->count;
#endif
```
and so we have removed a known bug.

Let's just wait for the tests to pass, and I'll squash the commits and merge.
